### PR TITLE
Support for MinIO Disco

### PR DESCRIPTION
**Pull request by dvaldivia**
Adds support for MinIO Disco

it defaults the start command for the statefulsets to the structure `instance-{0...X}.statefulset.minio.local`, this can be overriden with the argument `--hosts-template` for example `--hosts-template={{.StatefulSet}}-{{.Ellipsis}}.{{.HLService}}` would override this default.

Before creating a statefulset, we probe if `MinIO Disco` is configured globally, in case it's not, we get the IP of the `minio-disco` service and configure the statefulset to have a custom `dnsPolicy`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,6 +31,72 @@ To start MinIO-Operator with default configuration, use the `minio-operator.yaml
 kubectl apply -f https://raw.githubusercontent.com/minio/minio-operator/master/minio-operator.yaml
 ```
 
+MinIO Operator relies on [MinIO Disco](https://github.com/minio/disco) for Service Discovery and so that applications inside the cluster can discover instances if they lookup an instance by bucket (i.e: `bucket1.tenant.minio.local`), so you need to forward any request for the domain `minio.local` on you cluster DNS.
```
Comment (kerneltime) [inline README.md:34]: This implies that tenant name needs to be unique across all namespaces. Would we need to add a corresponding check in m3 or in the operator itself? This is with the model the CRD lives in the namespace of the minio instance.

Comment (dvaldivia) [reply README.md:34]: You are right, I’ll update it so it includes the namespace in the route

File: minio-operator.yaml [hunk 2]
```diff
@@ -90,6 +90,101 @@ spec:
...
+  template:
+    metadata:
+      labels:
+        app: minio-disco
+    spec:
+      containers:
+        - args:
+            - /disco
+            - server
+          image: minio/disco:latest
```
Comment (harshavardhana) [inline minio-operator.yaml:176]:
```suggestion
          image: minio/disco:v0.0.1
```

File: minio-operator.yaml [hunk 3]
```diff
@@ -90,6 +90,101 @@ spec:
...
+          imagePullPolicy: Always
```
Comment (harshavardhana) [inline minio-operator.yaml:177]:
```suggestion
          imagePullPolicy: IfNotPresent
```

File: pkg/apis/operator.min.io/v1/helper.go [hunk 4]
```diff
@@ -253,20 +256,28 @@ func (mi *MinIOInstance) TemplatedMinIOHosts(hostsTemplate string) []string {
 		if err = tmpl.Execute(output, data); err != nil {
 			continue
 		}
+		fmt.Println("output.String()")
```
Comment (harshavardhana) [inline pkg/apis/operator.min.io/v1/helper.go:259]: Thanks - but no thanks :-)

File: README.md [hunk 5]
```diff
@@ -31,6 +31,72 @@ To start MinIO-Operator with default configuration, use the `minio-operator.yaml
...
+    }
+    minio.local:53 {
+        errors
+        cache 30
+        forward . 10.109.234.52
+    }
+```
+Afterwards, restart the coredns pods on the `kube-system` namespace
+```bash
+$ kubectl -n kube-system delete pod $(kubectl -n kube-system get pods  | grep coredns | awk '{print $1}')
```
Comment (kerneltime) [inline README.md:97]: This is the idomatic way ```kubectl rollout restart -n kube-system deployment coredns```

File: go.sum [hunk 6]
```diff
@@ -181,8 +181,6 @@ github.com/hpcloud/tail v1.0.0 h1:nfCOvKYfkgYP8hkirhJocXT2+zOD8yUNjXaWfTlyFKI=
 github.com/hpcloud/tail v1.0.0/go.mod h1:ab1qPbhIpdTxEkNHXyeSf5vhxWSCs/tWer42PpOxQnU=
 github.com/imdario/mergo v0.3.5 h1:JboBksRwiiAJWvIYJVo46AfV+IAIKZpfrSzVKj42R4Q=
 github.com/imdario/mergo v0.3.5/go.mod h1:2EnlNZ0deacrJVfApfmtdGgDfMuh/nq6Ok1EcJh5FfA=
-github.com/imdario/mergo v0.3.6 h1:xTNEAn+kxVO7dTZGu0CegyqKZmoWFI0rF8UxjlB2d28=
```
Comment (kerneltime) [inline go.sum:184]: Why this change?

File: README.md (hunk 5 above)
Comment (kerneltime) [inline README.md:97]:
```suggestion
$ kubectl rollout restart -n kube-system deployment coredns
```

File: pkg/apis/operator.min.io/v1/helper.go [hunk 7]
```diff
@@ -470,3 +481,23 @@ func (mi *MinIOInstance) OwnerRef() []metav1.OwnerReference {
...
+func (mi *MinIOInstance) ProbeDisco(clientset kubernetes.Interface) (string, error) {
+	// if we get a response, it means disco is
+	// properly configured in the cluster, else we fallback to get the `minio-disco` service IP and configure
+	// that on the dnsPolicy for the pods of the statefulset
+	var discoSvcIP string
+	_, err := net.LookupIP(discoProbeDomain)
+	if err != nil {
+		discoSvc, err := clientset.
+			CoreV1().
+			Services(mi.Namespace).
```
Comment (kerneltime) [inline pkg/apis/operator.min.io/v1/helper.go:495]: Why look for the disco service in the namespace for the minio Instance?

File: pkg/resources/statefulsets/minio-statefulset.go [hunk 8]
```diff
@@ -390,5 +390,20 @@ func NewForMinIO(mi *miniov1.MinIOInstance, serviceName string, hostsTemplate st
 			ss.Spec.VolumeClaimTemplates = append(ss.Spec.VolumeClaimTemplates, pvClaim)
 		}
 	}
+
+	// configure the pod to register into disco
+	if ss.Spec.Template.ObjectMeta.Annotations == nil {
+		ss.Spec.Template.ObjectMeta.Annotations = map[string]string{}
+	}
+	ss.Spec.Template.ObjectMeta.Annotations["io.min.disco"] = fmt.Sprintf("{.metadata.name}.%s", mi.HostPostfix())
+	// if a Disco Service IP was provided, global service discovery is not enabled, so we are going to configure the
```
Comment (kerneltime) [inline pkg/resources/statefulsets/minio-statefulset.go:399]: This is to speed up Pod discovery, based on recent testing by @harshavardhana this is not needed. Also, that should be a separate PR.

File: go.sum (hunk 6 above)
Comment (dvaldivia) [reply go.sum:184]: I just ran `go mod tidy`

File: pkg/resources/statefulsets/minio-statefulset.go (hunk 8 above)
Comment (dvaldivia) [reply pkg/resources/statefulsets/minio-statefulset.go:399]: what do you mean it's not needed? the fallback is needed to guarantee the service discovery always works for MinIO even if it has not been configured globally

File: pkg/apis/operator.min.io/v1/helper.go (hunk 4 above)
Comment (dvaldivia) [reply pkg/apis/operator.min.io/v1/helper.go:259]: hahaha I thought I had removed all these, fixed.

File: README.md (hunk 1 above)
Comment (dvaldivia) [reply README.md:34]: done

File: pkg/resources/statefulsets/minio-statefulset.go (hunk 8 above)
Comment (kerneltime) [reply pkg/resources/statefulsets/minio-statefulset.go:399]: As far as I know, nodes currently talk to each other using node names which is resolved using standard dns configuration for stateful set node names. Minio is not aware of how it is presented as a single service.

Comment (kerneltime) [reply pkg/resources/statefulsets/minio-statefulset.go:399]: This will allow minio pods to resolve the Bucket DNS names? Is there another purpose for it?

File: pkg/controller/cluster/main-controller.go [hunk 9]
```diff
@@ -489,8 +496,14 @@ func (c *Controller) syncHandler(key string) error {
 			if err != nil {
 				return err
 			}
+
+			// Probe if disco is configured at cluster level
+			discoSvcIP, err := mi.ProbeDisco(c.kubeClientSet)
+			if err != nil {
+				return err
+			}
```
Comment (nitisht) [inline pkg/controller/cluster/main-controller.go:504]: Do we need to probe here again, since this is the zone addition case meaning probing must have happened already

File: README.md [hunk 10]
```diff
@@ -31,6 +31,72 @@ To start MinIO-Operator with default configuration, use the `minio-operator.yaml
...
+$ kubectl get svc minio-disco -o wide
+NAME          TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)         AGE   SELECTOR
+minio-disco   ClusterIP   10.109.234.52   <none>        53/UDP,53/TCP   12m   app=minio-disco
+
+```
+
+Here we can see the IP is `10.109.234.52` so we are going to add that to the `Corefile` stored in the `coredns` configmap inside the `kube-system` namespace.
+
+```bash
+$ kubectl -n kube-system edit configmap corends
```
Comment (nitisht) [inline README.md:48]:
```suggestion
$ kubectl -n kube-system edit configmap coredns
```
